python/ql/lib/semmle/python/dataflow/new/internal/TypeTrackingImpl.qll

```diff
@@ -97,24 +97,25 @@ private module SummaryTypeTrackerInput implements SummaryTypeTracker::Input {
 
 private module TypeTrackerSummaryFlow = SummaryTypeTracker::SummaryFlow<SummaryTypeTrackerInput>;
 
-/**
- * Gets the name of a possible piece of content. For Python, this is currently only attribute names,
- * using the name of the attribute for the corresponding content.
- */
-private string getPossibleContentName() {
-  Stages::TypeTracking::ref() and // the TypeTracking::append() etc. predicates that we want to cache depend on this predicate, so we can place the `ref()` call here to get around identical files.
-  result = any(DataFlowPublic::AttrRef a).getAttributeName()
-}
-
 module TypeTrackingInput implements Shared::TypeTrackingInput {
   class Node = DataFlowPublic::Node;
 
   class LocalSourceNode = DataFlowPublic::LocalSourceNode;
 
-  class Content instanceof string {
-    Content() { this = getPossibleContentName() }
-
-    string toString() { result = this }
+  class Content extends DataFlowPublic::Content {
+    Content() {
+      // TODO: for now, it's not 100% clear if should support non-precise content in
+      // type-tracking, or if it will lead to bad results. We start with only allowing
+      // precise content, which should always be a good improvement! It also simplifies
+      // the process of examining new results from non-precise content steps in the
+      // future, since you will _only_ have to look over the results from the new
+      // non-precise steps.
+      this instanceof DataFlowPublic::AttributeContent
+      or
+      this instanceof DataFlowPublic::DictionaryElementContent
+      or
+      this instanceof DataFlowPublic::TupleElementContent
+    }
```

Comment:
I wonder if it is better to define a subset of `Content` called `PreciseContent` and refer to it here. I guess it depends on whether we expect to update content and want to remember also updating the list of precise content before we start editing this definition to not refer to precise content anymore. And on whether the concept of precise content might have value on its own...

Reply:
Also, did you consider using `ContentSet` here like Ruby does (and like we do above for `SummaryTypeTracker::Input`)? It seems like, since we are in the area, we might as well be future proof...

Reply:
If we start defining the same set of "precise content" again in the future, I'm all for it :+1:

I think it would be nice to move to `ContentSet like Ruby, but let's keep that to an other PR :+1: